_posts/2016-05-13-a-comprehensive-guide-to-interacting-with-imap-using-ruby.markdown

```diff
@@ -0,0 +1,260 @@
+
+```ruby
+[2] pry(main)> imap = Net::IMAP.new("imap.googlemail.com", 993, true, nil, false)
+=> #<Net::IMAP:0x007fa1c77064b0 ...
+```
+
+which skips SSL verification. Note that while this is probably not such a good
+idea, even the most popular Gmail Ruby gem,
+[Gmail](https://github.com/gmailgem/gmail/blob/353ddcc8cc0c5b57ad1d3a412f11365ccc12b7d6/lib/gmail/client/base.rb#L26),
+does this.
+
+After the connection is established, you can login using a couple of different
+ways. First, the most common way, using the email and the password:
+
+```ruby
+[3] pry(main)> imap.login(email, password)
+=> #<struct Net::IMAP::TaggedResponse
+ name="OK",
+ raw_data="RUBY0001 OK c3p0@ombulabs.com authenticated (Success)\r\n">
+```
+
+Second, you can include the [gmail_xoauth](https://github.com/nfo/gmail_xoauth)
+gem, which adds the XOAUTH2 authenticator to `Net::IMAP`, allowing you to use an
+OAuth 2.0 token to login:
+
+```ruby
+[3] pry(main)> imap.authenticate("LOGIN", email, your_oauth2_token)
+=> #<struct Net::IMAP::TaggedResponse ...
+```
+
+## Listing available mailboxes
+
+To list all of your mailboxes:
+
+```ruby
+[4] pry(main)> imap.list("", "*")
+=> [#<struct Net::IMAP::MailboxList attr=[:All, :Hasnochildren], delim="/", name="[Gmail]/All Mail">,
+ #<struct Net::IMAP::MailboxList attr=[:Drafts, :Hasnochildren], delim="/", name="[Gmail]/Drafts">,
+ #<struct Net::IMAP::MailboxList attr=[:Hasnochildren, :Important], delim="/", name="[Gmail]/Important">,
+ #<struct Net::IMAP::MailboxList attr=[:Hasnochildren, :Sent], delim="/", name="[Gmail]/Sent Mail">,
+ #<struct Net::IMAP::MailboxList attr=[:Hasnochildren, :Junk], delim="/", name="[Gmail]/Spam">,
+ #<struct Net::IMAP::MailboxList attr=[:Flagged, :Hasnochildren], delim="/", name="[Gmail]/Starred">,
+ #<struct Net::IMAP::MailboxList attr=[:Hasnochildren, :Trash], delim="/", name="[Gmail]/Trash">]
+```
+
+To search your mailboxes by name, using `*` as a wildcard character:
+
+```ruby
+[5] pry(main)> imap.list("", "*Mail")
+=> [#<struct Net::IMAP::MailboxList attr=[:All, :Hasnochildren], delim="/", name="[Gmail]/All Mail">,
+ #<struct Net::IMAP::MailboxList attr=[:Hasnochildren, :Sent], delim="/", name="[Gmail]/Sent Mail">]
+```
+
+## Selecting a mailbox
+
+Most IMAP interactions require having selected a mailbox before doing them.
+There are two ways to do so:
+
+Read-only:
+
+```ruby
+[6] pry(main)> imap.examine("[Gmail]/All Mail")
+=> #<struct Net::IMAP::TaggedResponse
+ name="OK",
+ raw_data="RUBY0014 OK [READ-ONLY] [Gmail]/All Mail selected. (Success)\r\n">
+```
+
+Read and write:
+
+```ruby
+[7] pry(main)> imap.select("[Gmail]/All Mail")
+=> #<struct Net::IMAP::TaggedResponse
+ name="OK",
+ raw_data="RUBY0015 OK [READ-WRITE] [Gmail]/All Mail selected. (Success)\r\n">
+```
+
+## Searching email messages
+
+After selecting an email box, you can search email messages and return either
+the sequence number (`seqno`) for the email, or the unique id (`uid`).
```

Comment:
@mauro-oto it should say "the sequence number (`seqno`) or unique id (`uid`) for the email."